### PR TITLE
Added ability to remove user from CRM

### PR DESCRIPTION
**Pull request by RemcoSimonides**
To do in issue #3592 

"we need to add a delete button (+ confirmation dialog) on users. The delete should delete the user and all corresponding entry in the db (id in org userIds, invitations, notifications,...)"

### CODE REVIEW

**Comment by RemcoSimonides** [comment]
- [ ] Navigate user to previous page after deleting the user?

File: apps/backend-functions/src/users.ts [hunk 1]
```diff
@@ -148,8 +148,49 @@ export async function onUserDelete(
...
+  
+  // remove all notifications related to user
+  const notificationsRef = db.collection(`notifications`).where('toUserId', '==', user.uid);
+  const notificationsSnap = await notificationsRef.get();
+  notificationsSnap.forEach(notification => db.doc(`notifications/${notification.id}`).delete());
+
+  // delete blockframesAdmin doc?
+
+  // delete events owned by the user?
+
```
Comment (pldespaigne) [inline apps/backend-functions/src/users.ts:193]: maybe you also need to delete every files in the firebase storage under the path `users/<USER_ID>/` such as avatar & watermark

File: apps/backend-functions/src/users.ts [hunk 2]
```diff
@@ -148,8 +148,54 @@ export async function onUserDelete(
...
+  // remove permissions
+  if (!!user.orgId) {
+    const permissionsRef = db.doc(`permissions/${user.orgId}`);
+    const permissionsDoc = await permissionsRef.get();
+    const permissions = permissionsDoc.data() as PermissionsDocument;
+    const roles = permissions.roles;
+    delete roles[user.uid];
+    permissionsRef.update({roles}); 
+  }
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:176]: You cannot use existing `removeMember` (libs/organization/src/lib/organization/+state/organization.service.ts) for this ?

File: apps/backend-functions/src/users.ts [hunk 3]
```diff
@@ -148,8 +148,54 @@ export async function onUserDelete(
...
+
+  // remove all invitations related to user
+  const invitations = await getCollection<InvitationDocument>(`invitations`)
+  invitations
+    .filter(invitation => invitation.fromUser?.uid === user.uid || invitation.toUser?.uid === user.uid)
+    .map(invitation => invitation.id)
+    .map(id => db.doc(`invitations/${id}`).delete());
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:183]: You can move this to InvitationService

File: apps/backend-functions/src/users.ts [hunk 4]
```diff
@@ -148,8 +148,54 @@ export async function onUserDelete(
...
+  
+  // remove all notifications related to user
+  const notificationsRef = db.collection(`notifications`).where('toUserId', '==', user.uid);
+  const notificationsSnap = await notificationsRef.get();
+  notificationsSnap.forEach(notification => db.doc(`notifications/${notification.id}`).delete());
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:188]: You can move this to NotificationService

File: apps/backend-functions/src/users.ts [hunk 5]
```diff
@@ -148,8 +148,54 @@ export async function onUserDelete(
...
+
+  // delete blockframesAdmin doc?
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:190]: yes

File: apps/backend-functions/src/users.ts [hunk 6]
```diff
@@ -148,8 +148,54 @@ export async function onUserDelete(
...
+
+  // delete events owned by the user?
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:192]: There should not be any events directly linked to user currently (only orgId).

File: apps/backend-functions/src/users.ts (hunk 3 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/users.ts:183]: Maybe add a warning before deleting ?  Saying that all invitation will be removed

File: apps/backend-functions/src/users.ts (hunk 6 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/users.ts:192]: But (for the future) maybe add a warning before deleting ?  Saying that all event will be removed

File: libs/admin/src/lib/admin-panel/pages/user/user.component.ts [hunk 7]
```diff
@@ -100,4 +105,23 @@ export class UserComponent implements OnInit {
     await this.adminService.sendPasswordResetEmail(this.user.email);
     this.snackBar.open(`Reset password email sent to : ${this.user.email}`, 'close', { duration: 2000 });
   }
+
+  public async deleteUser() {
+    // check if user is not last superAdmin
+    const permissions = await this.permissionService.getValue(this.user.orgId);
+    if (!!this.user.orgId && !this.permissionService.hasLastSuperAdmin(permissions, this.userId, 'member')) {
+      throw new Error('There must be at least one Super Admin in the organization.');
+    }
```
Comment (phpgeekfr) [inline libs/admin/src/lib/admin-panel/pages/user/user.component.ts:114]: This is already checked in `removeMember` maybe you can update it a bit so it fit your needs

**Review by phpgeekfr — APPROVED** [review]
LGTM but look at comments

File: apps/backend-functions/src/users.ts (hunk 2 above)
Comment (RemcoSimonides) [reply apps/backend-functions/src/users.ts:176]: It was used as an inspiration ;)

File: libs/admin/src/lib/admin-panel/pages/user/user.component.ts [hunk 8]
```diff
@@ -132,4 +138,25 @@ export class UserComponent implements OnInit {
...
+
+    this.dialog.open(ConfirmComponent, {
+      data: {
+        title: 'Are you sure you want to delete this user?',
+        question: 'this action is irreversable',
```
Comment (phpgeekfr) [inline libs/admin/src/lib/admin-panel/pages/user/user.component.ts:152]: Typo on irreversable ?

**Review by phpgeekfr — APPROVED** [review]

File: apps/backend-functions/src/users.ts [hunk 9]
```diff
@@ -148,8 +148,52 @@ export async function onUserDelete(
...
+    .filter(invitation => invitation.fromUser?.uid === user.uid || invitation.toUser?.uid === user.uid)
+    .map(invitation => invitation.id)
+    .map(id => db.doc(`invitations/${id}`).delete());
+  
+  // remove all notifications related to user
+  const notificationsRef = db.collection(`notifications`).where('toUserId', '==', user.uid);
+  const notificationsSnap = await notificationsRef.get();
+  notificationsSnap.forEach(notification => db.doc(`notifications/${notification.id}`).delete());
+
+  // delete blockframesAdmin doc?
```
Comment (phpgeekfr) [inline apps/backend-functions/src/users.ts:190]: you can remove "?"

**Pull request merged**